# Project 2 - DL
## Image Classification with CNN
-------------------------------
Jessica Barón Martínez Código: 200924758

Nelson Andrés Rozo Cruz Código: 201728031

Wilfredo David Vega Buelvas, código 201727430

Camilo Torres Ovalle, código 201747423

1. Direct prediction
Download the dataset birds from http://www-cvr.ai.uiuc.edu/ponce_grp/data/.
Use Keras and a CNN from Keras Applications pretrained on ImageNet, to classify the images in the birds dataset. Construct a confusion matrix that relates the bird classes with the 10 most frequent classes from ImageNet predicted by the model.
Discuss the results.


In [1]:
import pandas as pd
import os
import numpy as np
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.multiclass import OneVsRestClassifier
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.metrics import r2_score, roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
from tqdm import tqdm
import os
import os
from skimage import io
from keras.preprocessing import image
from keras.applications.imagenet_utils import preprocess_input, decode_predictions
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras.utils import np_utils
from sklearn.metrics import confusion_matrix
from array import array
%matplotlib inline

C:\Users\jessi\Anaconda3\lib\site-packages\h5py\__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


# Read Images

In [2]:
path ='../proyectofinal2/birds'

In [3]:
tipo_1=os.listdir(os.path.join(path, 'egret'))

egret = []
for img in tipo_1:
    img_path=os.path.join(path, 'egret', img)
    img = image.load_img(img_path, target_size=(32,32))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    egret.append(x)



In [4]:
tipo_2=os.listdir(os.path.join(path, 'mandarin'))

mandarin = []
for img in tipo_2:
    img_path=os.path.join(path, 'mandarin', img)
    img = image.load_img(img_path, target_size=(32,32))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    mandarin.append(x)
#mandarin= np.concatenate(mandarin)

In [5]:
tipo_3=os.listdir(os.path.join(path, 'owl'))

owl= []
for img in tipo_3:
    img_path=os.path.join(path, 'owl', img)
    img = image.load_img(img_path, target_size=(32,32))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    owl.append(x) 

In [6]:
tipo_4=os.listdir(os.path.join(path, 'puffin'))

puffin= []
for img in tipo_4:
    img_path=os.path.join(path, 'puffin', img)
    img = image.load_img(img_path, target_size=(32,32))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    puffin.append(x)  

In [7]:
tipo_5=os.listdir(os.path.join(path, 'toucan'))

toucan= []
for img in tipo_5:
    img_path=os.path.join(path, 'toucan', img)
    img = image.load_img(img_path, target_size=(32,32))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    toucan.append(x)   

In [8]:
tipo_6=os.listdir(os.path.join(path, 'wood_duck'))

wood_duck= []
for img in tipo_6:
    img_path=os.path.join(path, 'wood_duck', img)
    img = image.load_img(img_path, target_size=(32,32))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    wood_duck.append(x)  

C:\Users\jessi\Anaconda3\lib\site-packages\PIL\TiffImagePlugin.py:692: UserWarning: Possibly corrupt EXIF data.  Expecting to read 80000 bytes but only got 0. Skipping tag 64640
  "Skipping tag %s" % (size, len(data), tag))
C:\Users\jessi\Anaconda3\lib\site-packages\PIL\TiffImagePlugin.py:692: UserWarning: Possibly corrupt EXIF data.  Expecting to read 2652897280 bytes but only got 0. Skipping tag 2
  "Skipping tag %s" % (size, len(data), tag))
C:\Users\jessi\Anaconda3\lib\site-packages\PIL\TiffImagePlugin.py:692: UserWarning: Possibly corrupt EXIF data.  Expecting to read 2656043008 bytes but only got 0. Skipping tag 2
  "Skipping tag %s" % (size, len(data), tag))
C:\Users\jessi\Anaconda3\lib\site-packages\PIL\TiffImagePlugin.py:692: UserWarning: Possibly corrupt EXIF data.  Expecting to read 65536 bytes but only got 0. Skipping tag 3
  "Skipping tag %s" % (size, len(data), tag))
C:\Users\jessi\Anaconda3\lib\site-packages\PIL\TiffImagePlugin.py:692: UserWarning: Possibly corrupt EXIF 

In [9]:
datos= egret+ mandarin + owl+ puffin+ toucan+ wood_duck
datos = np.vstack( datos )

In [10]:
datos.shape

(600, 32, 32, 3)

In [11]:
y = ([0]*100+[1]*100+[2]*100+[3]*100+[4]*100+[5]*100)
y = np.vstack( y )

In [12]:
y.shape

(600, 1)

In [13]:
X_train, X_test, y_train, y_test = train_test_split(datos, y , test_size=0.2, random_state=42)

In [14]:
X_train.shape

(480, 32, 32, 3)

In [15]:
nb_classes = y_train.max() + 1
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)
X_train = X_train.astype("float32")
X_test = X_test.astype("float32")
X_train /= 255
X_test /= 255

## 2. Transferencia de aprendizaje 

In [16]:
from keras.applications.vgg16 import VGG16
model_vgg = VGG16(weights='imagenet', include_top=False)

In [17]:
xtrainvgg=model_vgg.predict(X_train)

In [18]:
xtrainvgg.shape

(480, 1, 1, 512)

In [19]:
xtestvgg=model_vgg.predict(X_test)

In [20]:
model = Sequential()
model.add(Flatten(input_shape=xtrainvgg.shape[1:]))
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(6, activation='softmax'))

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_1 (Flatten)          (None, 512)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 256)               131328    
_________________________________________________________________
dropout_1 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 6)                 1542      
Total params: 132,870
Trainable params: 132,870
Non-trainable params: 0
_________________________________________________________________


In [21]:
model.compile(loss='sparse_categorical_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

In [22]:
model.fit(xtrainvgg, y_train, epochs=50,verbose=2)

Epoch 1/50
 - 0s - loss: 1.7910 - acc: 0.2667
Epoch 2/50
 - 0s - loss: 1.3990 - acc: 0.4688
Epoch 3/50
 - 0s - loss: 1.2386 - acc: 0.5229
Epoch 4/50
 - 0s - loss: 1.1169 - acc: 0.5854
Epoch 5/50
 - 0s - loss: 1.0498 - acc: 0.6229
Epoch 6/50
 - 0s - loss: 0.9444 - acc: 0.6667
Epoch 7/50
 - 0s - loss: 0.9283 - acc: 0.6750
Epoch 8/50
 - 0s - loss: 0.8648 - acc: 0.6833
Epoch 9/50
 - 0s - loss: 0.7769 - acc: 0.7312
Epoch 10/50
 - 0s - loss: 0.7890 - acc: 0.7292
Epoch 11/50
 - 0s - loss: 0.7490 - acc: 0.7396
Epoch 12/50
 - 0s - loss: 0.6941 - acc: 0.7604
Epoch 13/50
 - 0s - loss: 0.6632 - acc: 0.7729
Epoch 14/50
 - 0s - loss: 0.6258 - acc: 0.7792
Epoch 15/50
 - 0s - loss: 0.6034 - acc: 0.8063
Epoch 16/50
 - 0s - loss: 0.5475 - acc: 0.8271
Epoch 17/50
 - 0s - loss: 0.5371 - acc: 0.8229
Epoch 18/50
 - 0s - loss: 0.4912 - acc: 0.8521
Epoch 19/50
 - 0s - loss: 0.5037 - acc: 0.8292
Epoch 20/50
 - 0s - loss: 0.4670 - acc: 0.8542
Epoch 21/50
 - 0s - loss: 0.4497 - acc: 0.8500
Epoch 22/50
 - 0s - lo

In [23]:
# evaluate the model
scores = model.evaluate(xtestvgg, y_test)

120/120 [==============================] - ETA:  - 0s 532us/step


In [24]:
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


acc: 70.83%


In [25]:
y_pred=model.predict_classes(xtestvgg)

In [26]:
# Confusion matrix
df_confusion=confusion_matrix(y_test, y_pred)

In [27]:
df_confusion

array([[19,  1,  1,  0,  0,  2],
       [ 2, 12,  1,  1,  3,  1],
       [ 2,  2, 13,  0,  1,  1],
       [ 2,  1,  0, 12,  0,  2],
       [ 0,  1,  0,  1, 15,  0],
       [ 1,  6,  0,  0,  3, 14]])

## Conclusión Transfer Learning

Nuestro modelo  tuvo una preficción de 70.83%

## 3. AJUSTE FINO

In [32]:
from keras.applications.vgg16 import VGG16
model_vggG= VGG16(weights='imagenet', include_top=False)

for layer in model_vgg.layers:
    layer.trainable = True
 
# Check the trainable status of the individual layers
for layer in model_vgg.layers:
    print(layer, layer.trainable)

<keras.engine.input_layer.InputLayer object at 0x000001C9362AB9B0> True
<keras.layers.convolutional.Conv2D object at 0x000001C9362ABC50> True
<keras.layers.convolutional.Conv2D object at 0x000001C9362ABB70> True
<keras.layers.pooling.MaxPooling2D object at 0x000001C9362ABA58> True
<keras.layers.convolutional.Conv2D object at 0x000001C937980390> True
<keras.layers.convolutional.Conv2D object at 0x000001C937980C18> True
<keras.layers.pooling.MaxPooling2D object at 0x000001C9379A7C50> True
<keras.layers.convolutional.Conv2D object at 0x000001C9379B8A20> True
<keras.layers.convolutional.Conv2D object at 0x000001C9379CA390> True
<keras.layers.convolutional.Conv2D object at 0x000001C9379DDA90> True
<keras.layers.pooling.MaxPooling2D object at 0x000001C937A04CF8> True
<keras.layers.convolutional.Conv2D object at 0x000001C937A17AC8> True
<keras.layers.convolutional.Conv2D object at 0x000001C937A2A4E0> True
<keras.layers.convolutional.Conv2D object at 0x000001C937A3AFD0> True
<keras.layers.pool

In [33]:
model_vggG.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_3 (InputLayer)         (None, None, None, 3)     0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, None, None, 64)    1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, None, None, 64)    36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, None, None, 64)    0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, None, None, 128)   73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, None, None, 128)   147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, None, None, 128)   0         
__________

In [34]:
xtrainvggG=model_vgg.predict(X_train)

In [35]:
xtrainvggG.shape

(480, 1, 1, 512)

In [36]:
xtestvggG=model_vgg.predict(X_test)

In [37]:
model2 = Sequential()
model2.add(Flatten(input_shape=xtrainvgg.shape[1:]))
model2.add(Dense(256, activation='relu'))
model2.add(Dropout(0.5))
model2.add(Dense(6, activation='softmax'))

model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 512)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 256)               131328    
_________________________________________________________________
dropout_2 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 6)                 1542      
Total params: 132,870
Trainable params: 132,870
Non-trainable params: 0
_________________________________________________________________


In [38]:
model2.compile(loss='sparse_categorical_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

In [40]:
model2.fit(xtrainvggG, y_train, epochs=50,verbose=2)

Epoch 1/50
 - 0s - loss: 1.7883 - acc: 0.2562
Epoch 2/50
 - 0s - loss: 1.3962 - acc: 0.4646
Epoch 3/50
 - 0s - loss: 1.2127 - acc: 0.5542
Epoch 4/50
 - 0s - loss: 1.1364 - acc: 0.5792
Epoch 5/50
 - 0s - loss: 1.0438 - acc: 0.6208
Epoch 6/50
 - 0s - loss: 0.9846 - acc: 0.6396
Epoch 7/50
 - 0s - loss: 0.9041 - acc: 0.6937
Epoch 8/50
 - 0s - loss: 0.8383 - acc: 0.7167
Epoch 9/50
 - 0s - loss: 0.8231 - acc: 0.7083
Epoch 10/50
 - 0s - loss: 0.7778 - acc: 0.7208
Epoch 11/50
 - 0s - loss: 0.7215 - acc: 0.7667
Epoch 12/50
 - 0s - loss: 0.6840 - acc: 0.7562
Epoch 13/50
 - 0s - loss: 0.6636 - acc: 0.7792
Epoch 14/50
 - 0s - loss: 0.6317 - acc: 0.7896
Epoch 15/50
 - 0s - loss: 0.5998 - acc: 0.7958
Epoch 16/50
 - 0s - loss: 0.5436 - acc: 0.8417
Epoch 17/50
 - 0s - loss: 0.5482 - acc: 0.8167
Epoch 18/50
 - 0s - loss: 0.5237 - acc: 0.8396
Epoch 19/50
 - 0s - loss: 0.4654 - acc: 0.8604
Epoch 20/50
 - 0s - loss: 0.4392 - acc: 0.8583
Epoch 21/50
 - 0s - loss: 0.4558 - acc: 0.8604
Epoch 22/50
 - 0s - lo

In [41]:
# evaluate the model
scores2 = model2.evaluate(xtestvggG, y_test)

120/120 [==============================] - ETA:  - 0s 706us/step


In [42]:
print("\n%s: %.2f%%" % (model2.metrics_names[1], scores2[1]*100))


acc: 64.17%


In [43]:
y_predd=model2.predict_classes(xtestvggG)

In [44]:
# Confusion matrix
df_confusion1=confusion_matrix(y_test, y_predd)

In [45]:
df_confusion1

array([[15,  1,  1,  3,  1,  2],
       [ 1, 11,  1,  2,  2,  3],
       [ 1,  1, 10,  3,  1,  3],
       [ 1,  0,  0, 13,  0,  3],
       [ 0,  1,  0,  1, 15,  0],
       [ 1,  8,  0,  1,  1, 13]])

Nuestro modelo  tuvo una predicción de 64.17%


Para esta caso se evidencia que con la activaciòn de todas las capas, el modelo baja su poder de predicciòn
para el set de datos que se ha seleccionado de train y test.
A pesar de eso, la predicciòn de los dos modelos es aceptable.